imap/imapd.c

```diff
@@ -3451,89 +3532,72 @@ static void cmd_idle(char *tag)
             continue;
+
         /* Don't show if they're not shown at this level of login */
-        if (!(base_capabilities[i].mask & flags))
+        if (!(mask & flags) || (imapd_authstate && !(mask & CAPA_POSTAUTH)))
             continue;
-        /* cheap and nasty version of LITERAL- (RFC 7888) support - just say so */
-        if (lminus && !strcmp(capa, "LITERAL+"))
-            capa = "LITERAL-";
-        /* print the capability */
-        if (need_space) prot_putc(' ', imapd_out);
-        else need_space = 1;
-        prot_printf(imapd_out, "%s", base_capabilities[i].str);
-    }
-
-    if (config_mupdate_server) {
-        prot_printf(imapd_out, " MUPDATE=mupdate://%s/", config_mupdate_server);
-    }
-
-    if (apns_enabled) {
-        prot_printf(imapd_out, " XAPPLEPUSHSERVICE");
-    }
 
-    if (tls_enabled() && !imapd_starttls_done && !imapd_authstate) {
-        prot_printf(imapd_out, " STARTTLS");
-    }
-    if (imapd_tls_required || imapd_authstate ||
-        (!imapd_starttls_done && (extprops_ssf < 2) &&
-         !config_getswitch(IMAPOPT_ALLOWPLAINTEXT))) {
-        prot_printf(imapd_out, " LOGINDISABLED");
-    }
-
-    /* add the SASL mechs */
-    if (!imapd_tls_required && (!imapd_authstate || saslprops.ssf) &&
-        sasl_listmech(imapd_saslconn, NULL,
-                      "AUTH=", " AUTH=",
-                      !imapd_authstate ? " SASL-IR" : "", &sasllist,
-                      NULL, &mechcount) == SASL_OK && mechcount > 0) {
-        prot_printf(imapd_out, " %s", sasllist);
-    } else {
-        /* else don't show anything */
-    }
+        /* Don't show if they're disabled by config switch */
+        if (mask & CAPA_CONFIG) {
+            enum imapopt config = base_capabilities[i].u.config;
+            if (!config_getswitch(config) == !(mask & CAPA_REVERSE))
+                continue;
+        }
 
-    if (!(flags & CAPA_POSTAUTH)) return;
+        /* Don't show if they're disallowed by server state */
+        if (mask & CAPA_STATE) {
+            int state = *base_capabilities[i].u.statep;
+            if (!state == !(mask & CAPA_REVERSE))
+                continue;
+        }
 
-    if (imapd_authstate && imapd_userisadmin) {
-        prot_printf(imapd_out, " UNAUTHENTICATE");
-    }
+        if (mask & CAPA_COMPLEX) {
+            /* Complex capability */
+            base_capabilities[i].u.complex();
+        }
+        else {
+            /* Print the capability */
+            const char **strp = NULL, *s = NULL, *valfmt = "";
+            int64_t *i64p = NULL, i64 = -1;
+            int num = 1, n = 0;
+
+            if (mask & CAPA_VALUE) {
+                /* Capability with formatted value */
+                strp = base_capabilities[i].u.value.strp;
+                i64p = base_capabilities[i].u.value.i64p;
+
+                /* Don't show if they're disabled by a NULL string argument */
+                if (strp && !(s = *strp))
+                    continue;
```

Comment:
It took a few minutes of hard staring to find where `s` was assigned for `CAPA_VALUE` (in the `if` condition)... I wonder if this could be more readable?

Though I started trying to suggest something, only to find it made it worse, so idk.

Reply:
I rewrote this (and updated the comments in struct capa_struct) to hopefully make it clearer